src/Microsoft.Health.Dicom.Core/Features/Workitem/WorkitemService.Cancel.cs

```diff
@@ -0,0 +1,134 @@
+    /// Provides functionality to process the list of <see cref="IDicomInstanceEntry"/>.
+    /// </summary>
+    public partial class WorkitemService
+    {
+        public async Task<CancelWorkitemResponse> ProcessCancelAsync(DicomDataset dataset, string workitemInstanceUid, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(dataset, nameof(dataset));
+
+            var workitemMetadata = await _workitemOrchestrator
+                .GetWorkitemMetadataAsync(workitemInstanceUid, cancellationToken)
+                .ConfigureAwait(false);
+
+            // Get the state transition result
+            var transitionStateResult = workitemMetadata?
+                .ProcedureStepState
+                .GetTransitionState(WorkitemStateEvents.NActionToRequestCancel);
+
+            if (ValidateCancelRequest(workitemInstanceUid, workitemMetadata, dataset, transitionStateResult))
+            {
+                // If there is a warning code, the workitem is already in the canceled state.
+                if (transitionStateResult.HasWarningWithCode)
+                {
+                    _responseBuilder.AddSuccess(
+                        string.Format(
+                            CultureInfo.InvariantCulture,
+                            DicomCoreResource.WorkitemIsInFinalState,
+                            workitemInstanceUid,
+                            workitemMetadata.ProcedureStepState,
+                            transitionStateResult.Code));
+                }
+                else
+                {
+                    await CancelWorkitemAsync(dataset, workitemMetadata, transitionStateResult.State, cancellationToken).ConfigureAwait(false);
+                }
+            }
+
+            return _responseBuilder.BuildCancelResponse();
+        }
+
+        private bool ValidateCancelRequest(string workitemInstanceUid, WorkitemMetadataStoreEntry workitemMetadata, DicomDataset dataset, WorkitemStateTransitionResult transitionStateResult)
+        {
+            try
+            {
+                GetValidator<CancelWorkitemDatasetValidator>().Validate(dataset);
+
+                CancelWorkitemDatasetValidator.ValidateProcedureStepStateInStore(workitemInstanceUid, workitemMetadata, transitionStateResult);
+
+                return true;
+            }
+            catch (Exception ex)
+            {
+                ushort? failureCode = FailureReasonCodes.ProcessingFailure;
+
+                switch (ex)
+                {
+                    case DatasetValidationException datasetValidationException:
+                        // should return 400
+                        failureCode = datasetValidationException.FailureCode;
+                        break;
+
+                    case DicomValidationException _:
+                    case ValidationException _:
+                        // should return 409
+                        failureCode = FailureReasonCodes.DatasetDoesNotMatchSOPClass;
```

Comment:
according to the spec: The request is inconsistent with the current state of the Target Workitem. For example, the Target Workitem is in the SCHEDULED or COMPLETED state.